torch/csrc/autograd/engine.cpp

```diff
@@ -809,35 +809,35 @@ void Engine::initialize_device_threads_pool() {
   std::call_once(start_device_threads_flag_, &Engine::start_device_threads, this);
 }
 
-void Engine::enqueue_blocked_task_on_cpu(NodeTask task) {
+std::shared_ptr<FutureVariableList> Engine::execute_graph_task_until_ready_queue_empty(
+    const std::shared_ptr<GraphTask>& graph_task,
+    std::shared_ptr<Node> root_to_execute,
+    bool incrementOutstandingTasks) {
   initialize_device_threads_pool();
-  std::shared_ptr<GraphTask> graph_task = task.base_.lock();
-  // The graph_task must be alive at this point, because internal autograd machinary
-  // who calls this API (Distributed Autograd Engine) increases outstanding_tasks_
-  // outside this API to keep the GraphTask alive
-  TORCH_INTERNAL_ASSERT(graph_task, "GraphTask is no longer valid!");
-  ready_queue(graph_task, at::kCPU)->push(
-      std::move(task), /* incrementOutstandingTasks */ false);
-}
-
-void Engine::execute_graph_task_with_continuation(const std::shared_ptr<GraphTask>& graph_task) {
   std::shared_ptr<ReadyQueue> graph_task_rq = graph_task->cpu_ready_queue_;
+  graph_task_rq->push(
+      NodeTask(graph_task, std::move(root_to_execute), InputBuffer(0)),
+      incrementOutstandingTasks);
+
+  set_device(CPU_DEVICE);
+  graph_task->owner_ = worker_device;
   while(!graph_task_rq->empty()) {
-    NodeTask task = graph_task_rq->pop();
     std::shared_ptr<GraphTask> local_graph_task;
-    if (!(local_graph_task = task.base_.lock())) {
-      continue;
-    }
     {
+      NodeTask task = graph_task_rq->pop();
+      if (!(local_graph_task = task.base_.lock())) {
+        continue;
+      }
       if (task.fn_ && !local_graph_task->has_error_.load()) {
         AutoGradMode grad_mode(local_graph_task->grad_mode_);
         try {
           evaluate_function(local_graph_task, task.fn_.get(), task.inputs_);
```

Comment:
we need to do continuation in evaluate_function? It seems it's not part of this PR.

Reply:
We certainly don't want to change local engine. 

I think what we want to avoid here is to get stuck silently when someone happens to do distributed autograd using GPU. Can we add an assertion here if node is expected to run on GPU?